src/core/ddsi/src/ddsi_typelookup.c

```diff
@@ -236,43 +50,59 @@ static struct writer *get_typelookup_writer (const struct ddsi_domaingv *gv, uin
   return wr;
 }
 
-bool ddsi_tl_request_type (struct ddsi_domaingv * const gv, const type_identifier_t *type_id)
+bool ddsi_tl_request_type (struct ddsi_domaingv * const gv, const ddsi_typeid_t *type_id, const ddsi_typeid_t ** dependent_type_ids, uint32_t dependent_type_id_count)
 {
-  ddsrt_mutex_lock (&gv->tl_admin_lock);
-  assert (!ddsi_typeid_none (type_id));
-  struct tl_meta *tlm = ddsi_tl_meta_lookup_locked (gv, type_id);
+  struct ddsi_typeid_str tidstr;
+  assert (ddsi_typeid_is_hash (type_id));
+  ddsrt_mutex_lock (&gv->typelib_lock);
+  struct ddsi_type *type = ddsi_type_lookup_locked (gv, type_id);
   GVTRACE ("tl-req ");
-  if (tlm->state != TL_META_NEW)
+  if (!type)
+  {
+    GVTRACE ("cannot find %s\n", ddsi_make_typeid_str (&tidstr, type_id));
+    ddsrt_mutex_unlock (&gv->typelib_lock);
+    return false;
+  }
+  else if (!dependent_type_id_count && (type->state == DDSI_TYPE_REQUESTED || type->xt.has_obj))
   {
     // type lookup is pending or the type is already resolved, so we'll return true
     // to indicate that the type request is done (or not required)
-    GVTRACE ("state not-new (%u) for "PTYPEIDFMT"\n", tlm->state, PTYPEID (*type_id));
-    ddsrt_mutex_unlock (&gv->tl_admin_lock);
+    GVTRACE ("state not-new for %s\n", ddsi_make_typeid_str (&tidstr, type_id));
+    ddsrt_mutex_unlock (&gv->typelib_lock);
     return true;
   }
 
   struct writer *wr = get_typelookup_writer (gv, NN_ENTITYID_TL_SVC_BUILTIN_REQUEST_WRITER);
   if (wr == NULL)
   {
     GVTRACE ("no pp found with tl request writer");
-    ddsrt_mutex_unlock (&gv->tl_admin_lock);
+    ddsrt_mutex_unlock (&gv->typelib_lock);
     return false;
   }
 
-  type_lookup_request_t request;
-  memcpy (&request.writer_guid, &wr->e.guid, sizeof (wr->e.guid));
-  request.sequence_number = ++tlm->request_seqno;
-  request.type_ids.n = 1;
-  request.type_ids.type_ids = &tlm->type_id;
+  DDS_Builtin_TypeLookup_Request request;
+  memset (&request, 0, sizeof (request));
+  memcpy (&request.header.requestId.writer_guid.guidPrefix, &wr->e.guid.prefix, sizeof (request.header.requestId.writer_guid.guidPrefix));
+  memcpy (&request.header.requestId.writer_guid.entityId, &wr->e.guid.entityid, sizeof (request.header.requestId.writer_guid.entityId));
+  type->request_seqno++;
+  request.header.requestId.sequence_number.high = (int32_t) (type->request_seqno >> 32);
+  request.header.requestId.sequence_number.low = (uint32_t) type->request_seqno;
+  // FIXME: request.header.instanceName = ...
```

Comment:
This one looks like it may need to be dealt with?

Reply:
This is added in ce4d331. The specs says it should be a 16-char string of the participant GUID, but the 16-byte GUID doesn't fit in so we'll include it as a 32-char string for now (an OMG xtypes spec issue is created for this: https://issues.omg.org/issues/DDSXTY14-53)